packages/angular_devkit/build_angular/src/utils/purge-cache.ts

```diff
@@ -0,0 +1,58 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import { BuilderContext } from '@angular-devkit/architect';
+import { PathLike, existsSync, promises as fsPromises } from 'fs';
+import { join } from 'path';
+import { normalizeCacheOptions } from './normalize-cache';
+
+/** Delete stale cache directories used by previous versions of build-angular. */
+export async function purgeStaleBuildCache(context: BuilderContext): Promise<void> {
+  const projectName = context.target?.project;
+  if (!projectName) {
+    return;
+  }
+
+  const metadata = await context.getProjectMetadata(projectName);
+  const { basePath, path, enabled } = normalizeCacheOptions(metadata, context.workspaceRoot);
+
+  if (!enabled || !existsSync(basePath)) {
+    return;
+  }
+
+  const entries = await fsPromises.readdir(basePath);
+
+  for (const entry of entries) {
+    const entryPath = join(basePath, entry);
+    if (entryPath === path) {
+      // Skip this directory since this is not stale.
+      continue;
+    }
+
+    // The below should be removed and replaced with just `rm` when support for Node.Js 12 is removed.
+    const { rm, rmdir } = fsPromises as typeof fsPromises & {
+      rm?: (
+        path: PathLike,
+        options?: {
+          force?: boolean;
+          maxRetries?: number;
+          recursive?: boolean;
+          retryDelay?: number;
+        },
+      ) => Promise<void>;
+    };
+
+    try {
+      if (rm) {
+        await rm(entryPath, { force: true, recursive: true, maxRetries: 3 });
+      } else {
+        await rmdir(entryPath, { recursive: true, maxRetries: 3 });
+      }
+    } catch {}
```

Comment:
Should we log or output the error somewhere? I'm concerned this might silently fail and we'll never know about it.

Reply:
Maybe a log message, since we don’t want this to be fatal.

Reply:
One thing to take into consideration here, is that there might be the case that 2 processes try to delete the same path at the same time which would cause an error. 

I think at this point, I prefer not to add an error log.

Reply:
PS: I added an E2E to cover the purging of stale cache.